contracts/governance/governance.cpp

```diff
@@ -0,0 +1,401 @@
+   system::detail::max_contract_size,       // proposal.upload_contract.bytecode
+   system::detail::max_contract_size,       // proposal.upload_contract.abi
+   system::detail::max_address_size,        // proposal.call_contract.contract_id
+   system::detail::max_argument_size,       // proposal.call_contract.args
+   system::detail::max_argument_size,       // proposal.set_system_call.target.system_call_bundle.contract_id
+   system::detail::max_address_size,        // proposal.set_system_contract.contract_id
+   system::detail::max_signatures_length,   // proposal.signatures length
+   system::detail::max_signature_size >;    // proposal.signatures
+
+bool proposal_updates_governance( const koinos::system::transaction& proposal )
+{
+   for ( uint64_t i = 0; i < proposal.get_operations().get_length(); i++ )
+   {
+      const auto& op = proposal.get_operations().get_const( i );
+      switch( op.get_which_op() )
+      {
+         case koinos::system::operation::FieldNumber::UPLOAD_CONTRACT:
+         {
+            const auto& upload_op = op.get_upload_contract();
+
+            std::string contract_id( reinterpret_cast< const char* >( upload_op.get_contract_id().get_const() ), upload_op.get_contract_id().get_length() );
+
+            // We're uploading the current governance contract
+            if ( contract_id == system::get_contract_id() )
+               return true;
+         }
+         break;
+
+         case koinos::system::operation::FieldNumber::CALL_CONTRACT:
+         break;
+
+         case koinos::system::operation::FieldNumber::SET_SYSTEM_CALL:
+         {
+            const auto& set_system_call_op = op.get_set_system_call();
+
+            // The governance contract is called during the pre_block_callback, so it is just as dangerous
+            // to allow for this system call to be overridden
+            if ( set_system_call_op.get_call_id().get() == std::underlying_type_t< koinos::chain::system_call_id >( koinos::chain::system_call_id::pre_block_callback ) )
+               return true;
+         }
+         break;
+
+         case koinos::system::operation::FieldNumber::SET_SYSTEM_CONTRACT:
+         {
+            const auto& set_system_contract_op = op.get_set_system_contract();
+
+            std::string contract_id( reinterpret_cast< const char* >( set_system_contract_op.get_contract_id().get_const() ), set_system_contract_op.get_contract_id().get_length() );
+
+            // We're setting the system status of the governance contract
+            if ( contract_id == system::get_contract_id() )
+               return true;
+         }
+         break;
+
+         default:
+         break;
+      }
+   }
+
+   return false;
+}
+
+submit_proposal_result submit_proposal( const submit_proposal_arguments& args )
+{
+   submit_proposal_result res;
+   res.set_value( false );
+
+   auto payer = koinos::system::get_transaction_field( "header.payer" );
```

Comment:
I propose to add the field `proposer` in `submit_proposal_arguments` and ask him to pay for the fee.
Then the unique role of `header.payer` is to pay for the mana.